src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -517,135 +168,287 @@ namespace cublog
 
-    for (; !finished ;)
+    for (bool finished = false; !finished ;)
       {
-	bool adding_finished = false;
-	redo_parallel::redo_job_base *const job = m_queue.pop_job (adding_finished);
+	pop_jobs (jobs_vec);
+	const bool jobs_popped { !jobs_vec->empty ()};
 	m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_POP);
-
-	if (job == nullptr && adding_finished)
+	if (!jobs_popped && m_adding_finished.load ())
 	  {
 	    finished = true;
 	  }
 	else
 	  {
-	    if (job == nullptr)
+	    if (!jobs_popped)
 	      {
 		// TODO: if needed, check if requested to finish ourselves
 
+		// no notification towards the parent for empty here, because at the next attempt
+		// we might get more work
+
 		// expecting more data, sleep and check again
 		std::this_thread::sleep_for (std::chrono::milliseconds (WAIT_AND_CHECK_MILLIS));
 		m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_SLEEP);
 	      }
 	    else
 	      {
 		THREAD_ENTRY *const thread_entry = &context;
-		job->execute (thread_entry, m_redo_context);
-		m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_EXECUTE);
-
-		m_queue.notify_job_finished (job);
-
-		job->retire (m_task_idx);
-		m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_RETIRE);
+		for (auto &job : *jobs_vec)
+		  {
+		    if (m_do_monitor_not_applied_log_lsa)
+		      {
+			set_not_applied_log_lsa_from_pop_side (job->get_log_lsa ());
+		      }
+		    job->execute (thread_entry, m_redo_context);
+		    job->retire (m_task_idx);
+		  }
+
+		// pointers still present in the vector are either:
+		//  - already passed on to the reusable job container
+		//  - dangling, as they have deleted themselves
+		jobs_vec->clear ();
+		m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_EXECUTE_AND_RETIRE);
 	      }
 	  }
       }
 
+    assert (jobs_vec->empty ());
+
     m_task_state_bookkeeping.set_inactive ();
+    m_in_execution.store (false);
   }
 
-  void redo_parallel::redo_task::retire ()
+  void
+  redo_parallel::redo_task::retire ()
   {
     // avoid self destruct, will be deleted by owning class
   }
 
-  void redo_parallel::redo_task::log_perf_stats () const
+  void
+  redo_parallel::redo_task::log_perf_stats () const
   {
     std::stringstream ss;
     ss << "Log recovery redo worker thread " << m_task_idx << " perf stats";
     m_perf_stats.log (ss.str ().c_str ());
   }
 
-  void redo_parallel::redo_task::accumulate_perf_stats (
+  void
+  redo_parallel::redo_task::accumulate_perf_stats (
 	  cubperf::stat_value *a_output_stats, std::size_t a_output_stats_size) const
   {
     m_perf_stats.accumulate (a_output_stats, a_output_stats_size);
   }
 
+  bool
+  redo_parallel::redo_task::is_idle () const
+  {
+    bool is_vector_empty = false;
+    {
+      std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
+      is_vector_empty = m_produce_vec->empty ();
+    }
+    return is_vector_empty && !m_in_execution.load ();
```

Comment:
Too complicated code for a simple check.
```suggestion
    std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
    return m_produce_vec->empty () && !m_in_execution.load ();
```